### PR TITLE
kyber: keep coefficients tangled     

### PR DESCRIPTION
**Pull request by bwesterb**
This keeps coefficients longer in an order convenient for AVX2-optimized
(Inv)NTT and MulHat.

### CODE REVIEW

File: pke/kyber/internal/common/generic.go [hunk 1]
```diff
@@ -39,6 +42,17 @@ func (p *Poly) InvNTT() {
...
+//
+// Requires a and b to be in "tangled" order, see Tangle().  p will be in
+// tangled order as well.
 func (p *Poly) MulHat(a, b *Poly) {
 	p.mulHatGeneric(a, b)
 }
+
+// Puts p into the right form to be used with (among others) InvNTT().
+func (p *Poly) Tangle() {
+}
```
Comment (armfazh) [inline pke/kyber/internal/common/generic.go:54]: empty function?

File: pke/kyber/internal/common/generic.go [hunk 2]
```diff
@@ -39,6 +42,17 @@ func (p *Poly) InvNTT() {
...
+
+// Puts p back into standard form.
+func (p *Poly) Detangle() {
+}
```
Comment (armfazh) [inline pke/kyber/internal/common/generic.go:58]: empty function?

File: pke/kyber/internal/common/amd64.go [hunk 3]
```diff
@@ -253,3 +259,17 @@ func (p *Poly) MulHat(a, b *Poly) {
 		p.mulHatGeneric(a, b)
 	}
 }
+
+// Puts p into the right form to be used with (among others) InvNTT().
+func (p *Poly) Tangle() {
+	if cpu.X86.HasAVX2 {
+		tangleAVX2((*[N]int16)(p))
+	}
```
Comment (armfazh) [inline pke/kyber/internal/common/amd64.go:267]: Non-AVX2 implementation assumes the data is in the right order?

**Review by armfazh — COMMENTED** [review]
When working with vectors, a lot of time is spent on moving data on the right position. Curious about the performance improvements at the top-level introduced by these changes.

File: pke/kyber/internal/common/generic.go (hunk 1 above)
Comment (bwesterb) [reply pke/kyber/internal/common/generic.go:54]: In generic code the tangled order is the standard one.  I'll add a comment.

**Comment by bwesterb** [comment]
> When working with vectors, a lot of time is spent on moving data on the right position. Curious about the performance improvements at the top-level introduced by these changes.

It doesn't look very dramatic, because current runtime is dominated by unoptimised SHAKE/SHA3 and sampling.

```
BenchmarkGenerateKey/Kyber512-4      30840         30692         -0.48%
BenchmarkGenerateKey/Kyber768-4      51408         51117         -0.57%
BenchmarkGenerateKey/Kyber1024-4     80796         80827         +0.04%
BenchmarkEncapsulate/Kyber512-4      21273         20879         -1.85%
BenchmarkEncapsulate/Kyber768-4      28699         26204         -8.69%
BenchmarkEncapsulate/Kyber1024-4     34527         33062         -4.24%
BenchmarkDecapsulate/Kyber512-4      24079         24085         +0.02%
BenchmarkDecapsulate/Kyber768-4      30750         29535         -3.95%
BenchmarkDecapsulate/Kyber1024-4     38752         37576         -3.03%
```

**Review by armfazh — APPROVED** [review]

**Pull request merged**